realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm;
+
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.Map;
+
+import io.realm.internal.ColumnIndices;
+import io.realm.internal.RealmProxyMediator;
+import io.realm.internal.SharedGroup;
+import io.realm.internal.log.RealmLog;
+
+/**
+ * To cache {@link Realm}, {@link DynamicRealm} instances and related resources.
+ * Every thread will share the same {@link Realm} and {@link DynamicRealm} instances which are referred to the same
+ * {@link RealmConfiguration}.
+ * One {@link RealmCache} is created for each {@link RealmConfiguration}.
+ */
+class RealmCache {
+
+    interface RealmCacheCallback {
+        void onResult(int count);
+    }
+
+    private static class RefAndCount {
+        // The Realm instance in this thread.
+        private final ThreadLocal<BaseRealm> localRealm = new ThreadLocal<BaseRealm>();
+        // How many references to this Realm instance in this thread.
+        private final ThreadLocal<Integer> localCount = new ThreadLocal<Integer>();
+        // How many threads have instances refer to this configuration.
+        private int globalCount = 0;
+    }
+    private enum RealmCacheType {
+        TYPED_REALM,
+        DYNAMIC_REALM;
+
+        static RealmCacheType valueOf(Class<? extends BaseRealm> cls) {
+            if (cls == Realm.class) {
+                return TYPED_REALM;
+            } else if (cls == DynamicRealm.class) {
+                return DYNAMIC_REALM;
+            }
+
+            throw new IllegalArgumentException(WRONG_REALM_CLASS_MESSAGE);
+        }
+    }
+    // Separated references and counters for typed Realm and dynamic Realm.
+    private final EnumMap<RealmCacheType, RefAndCount> refAndCountMap;
+
+    private RealmConfiguration configuration;
+    // Column indices are cached to speed up opening typed Realm. If a Realm instance is created in one thread, creating
+    // Realm instances in other threads doesn't have to initialize the column indices again.
+    private ColumnIndices typedColumnIndices;
+
+    // Different Realm configuration with same path are not allowed.
```

Comment:
The comment seems a bit odd, since it conflicts with the message in the class header. Given that description I would assume that this constraint is already enforced?